src/main/java/org/zalando/jackson/datatype/money/MonetaryAmountSerializer.java

```diff
@@ -12,20 +17,35 @@
 import java.io.IOException;
 import java.util.Locale;
 
-final class MonetaryAmountSerializer extends JsonSerializer<MonetaryAmount> {
+final class MonetaryAmountSerializer extends StdSerializer<MonetaryAmount> {
 
     private final FieldNames names;
     private final AmountWriter writer;
     private final MonetaryAmountFormatFactory factory;
 
     MonetaryAmountSerializer(final FieldNames names, final AmountWriter writer,
             final MonetaryAmountFormatFactory factory) {
+        super(MonetaryAmount.class);
         this.writer = writer;
         this.factory = factory;
         this.names = names;
     }
 
     @Override
+    public void acceptJsonFormatVisitor(JsonFormatVisitorWrapper visitor, JavaType typeHint) throws JsonMappingException {
+        JsonObjectFormatVisitor jsonObjectFormatVisitor = visitor.expectObjectFormat(typeHint);
+        JavaType numberType = visitor.getProvider().constructType(Number.class);
+        JavaType currencyUnitType = visitor.getProvider().constructType(CurrencyUnit.class);
+        JavaType stringType = visitor.getProvider().constructType(String.class);
+        JsonSerializer<Object> numberSerializer = visitor.getProvider().findValueSerializer(Number.class);
+        JsonSerializer<Object> currencyUnitSerializer = visitor.getProvider().findValueSerializer(CurrencyUnit.class);
+        JsonSerializer<Object> stringSerializer = visitor.getProvider().findValueSerializer(String.class);
+        jsonObjectFormatVisitor.property(names.getAmount(), numberSerializer, numberType);
```

Comment:
There is a feature to serialize amount as string which I believe is not respected here

https://github.com/zalando/jackson-datatype-money#serialization

Reply:
Right. I think we need to extract the type from the corresponding `AmountWriter`.

@aboivin The `AmountWriter` interface is package private, so you're free to change it. I'd suggest something like:

```java
interface AmountWriter<T> {
    Class<T> getType();
    T write(final MonetaryAmount amount);
}
```

Your current implementation could then use `writer.getType()` instead of `Number.class`.